### PR TITLE
lightning: support data files with bom header

### LINKED ISSUE #40744: Lightning cannot Import CSV files with UTF-8 BOM
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. Generate a CSV file and the schema file.  In the CSV file, the head of the file contains a UTF-8 BOM (0xEF, 0xBB, 0xBF).
Here's an example of the CSV file: 
```csv
"id","val1"
1,"aaa01"
2,"aaa01"
3,"aaa02"
4,"aaa02"
5,"aaa05"
```

Here's the binary representation of the file:
```bash
$ xxd src-data/bom-data/mytest.testtbl.csv
# result: 
# 00000000: efbb bf22 6964 222c 2276 616c 3122 0a31  ..."id","val1".1
# 00000010: 2c22 6161 6130 3122 0a32 2c22 6161 6130  ,"aaa01".2,"aaa0
# 00000020: 3122 0a33 2c22 6161 6130 3222 0a34 2c22  1".3,"aaa02".4,"
# 00000030: 6161 6130 3222 0a35 2c22 6161 6130 3522  aaa02".5,"aaa05"
# 00000040: 0a
```

3. Import data using Lightning

### 2. What did you expect to see? (Required)
The import should succeed

### 3. What did you see instead (Required)
The import failed.  Failure message: 
tidb lightning encountered error: [Lightning:Restore:ErrEncodeKV]encode kv error in file mytest.testtbl.csv:0 at offset 4: syntax error: cannot have consecutive fields without separator

### 4. What is your TiDB version? (Required)

``` plain
Release Version: v6.5.0
Edition: Community
Git Commit Hash: 706c3fa3c526cdba5b3e9f066b1a568fb96c56e3
Git Branch: heads/refs/tags/v6.5.0
UTC Build Time: 2022-12-27 03:42:38
GoVersion: go1.19.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv
```

### PR DESCRIPTION
**Pull request by dsdashun**
### What problem does this PR solve?

Issue Number: close #40744 

Problem Summary:

### What is changed and how it works?
When reading the CSV / SQL data file's first block, trim the BOM header, and update the processed position.
Originally I planned to use a wrapped Reader to handle that.  However, this will cause the processed position in parser not updated, thus resulting in abnormalities.  
For example, when the file has a BOM header, after processing the first line data with 10 bytes, the parser position should be set to 10 + 3 = 13 , not 10.  If set to 10, after processing all the file data, the final position is not equal to the chunk end offset, which is calculated by the file size.  Then this termination if clause won't be hit: https://github.com/pingcap/tidb/blob/4686338d0eef4135b41a7f4a5a70ee80ff267040/br/pkg/lightning/restore/restore.go#L2722-L2725 .  That will  cause the parser re-scan some file data again.  
So the data manipulation is handled when reading data blocks.  In this way we can also update the processed positions in the block parser.  

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/tests/lightning_bom_file/data/mytest.testtbl.csv [hunk 1]
```diff
@@ -0,0 +1,6 @@
+﻿id,val1
```
Comment (lance6716) [inline br/tests/lightning_bom_file/data/mytest.testtbl.csv:1]: Maybe check the BOM really exists in run.sh 😂

**Review by lance6716 — APPROVED** [review]

File: br/tests/lightning_bom_file/data/mytest.testtbl.csv (hunk 1 above)
Comment (buchuitoudegou) [reply br/tests/lightning_bom_file/data/mytest.testtbl.csv:1]: I remember the BOM header cannot be presented properly in the text editor...Maybe it's just ignored similarly by GitHub

Comment (lance6716) [reply br/tests/lightning_bom_file/data/mytest.testtbl.csv:1]: it has no visual output, but we can use `xxd` or something tools to check it. I'm afraid that the BOM header will be deleted accidentally in future.

Comment (dsdashun) [reply br/tests/lightning_bom_file/data/mytest.testtbl.csv:1]: addressed in 97497035d8e33f955affbabfd2fb69f6d1621263

**Comment by dsdashun** [comment]
integration BR test passed: https://ci.pingcap.net/blue/organizations/jenkins/br_ghpr_unit_and_integration_test/detail/br_ghpr_unit_and_integration_test/14871/pipeline

**Pull request merged**